core/domain/user_id_migration.py

```diff
@@ -0,0 +1,515 @@
+            if 'BaseSnapshotMetadataModel' in base_classes:
+                continue
+            # BaseSnapshotContentModel and models that inherit from it
+            # are checked from the associated VersionedModel.
+            if 'BaseSnapshotContentModel' in base_classes:
+                continue
+            if (model_class.get_deletion_policy() ==
+                    base_models.DELETION_POLICY.NOT_APPLICABLE):
+                continue
+            if model_class.has_reference_to_user_id(gae_id):
+                yield ('FAILURE - HAS REFERENCE TO GAE ID',
+                       (gae_id, model_class.__name__))
+                success = False
+
+        if success:
+            yield ('SUCCESS', (gae_id, user_model.id))
+
+    @staticmethod
+    def reduce(key, status):
+        """Implements the reduce function for this job."""
+        yield (key, status)
+
+
+class ModelsUserIdsHaveUserSettingsVerificationJob(
+        jobs.BaseMapReduceOneOffJobManager):
+    """One-off job for going through all the models that contain user IDs. This
+    job checks that all the user IDs used in the model have their corresponding
+    UserSettingsModel defined.
+    """
+
+    @staticmethod
+    def _does_user_settings_model_exits(user_id):
+        """Check if UserSettingsModel exists for the user_id.
+
+        Args:
+            user_id: str. User ID that should have its UserSettingsModel.
+
+        Returns:
+            True if UserSettingsModel with id equal to user_id exists, False
+            otherwise.
+        """
+        if user_id == feconf.SYSTEM_COMMITTER_ID:
+            return True
+        return user_models.UserSettingsModel.get_by_id(user_id) is not None
+
+    @staticmethod
+    def _check_id_and_user_id_exist(model_id, user_id):
+        """Check if UserSettingsModel exists for user_id and model id contains
+        user_id.
+
+        Args:
+            model_id: str. ID of the model that should contain the user_id.
+            user_id: str. User ID that should have its UserSettingsModel.
+
+        Returns:
+            True if UserSettingsModel with id as user_id in model exists, False
+            otherwise.
+        """
+        if user_id not in model_id:
+            return False
+        if user_id == feconf.SYSTEM_COMMITTER_ID:
+            return True
+        return user_models.UserSettingsModel.get_by_id(user_id) is not None
+
+    @staticmethod
+    def _check_one_field_exists(model):
+        """Check if UserSettingsModel exists for one field.
+
+        Args:
+            model: instance. Model being checked.
+
+        Returns:
+            True if UserSettingsModel with id as user ID field in model exists,
+            False otherwise.
+        """
+        model_class = model.__class__
+        model_values = model.to_dict()
+        user_id = model_values[model_class.get_user_id_migration_field()._name]  # pylint: disable=protected-access
+        if user_id == feconf.SYSTEM_COMMITTER_ID:
```

Comment:
Ditto -- why can't this be absorbed in the get_by_id "return" below?

Reply:
Replied in prev comment.